xds/src/test/java/io/grpc/xds/XdsSdsClientServerTest.java

```diff
@@ -0,0 +1,224 @@
+/*
+ * Copyright 2019 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.xds;
+
+import static com.google.common.truth.Truth.assertThat;
+import static io.grpc.xds.XdsClientWrapperForServerSdsTest.buildFilterChainMatch;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+
+import io.envoyproxy.envoy.api.v2.auth.DownstreamTlsContext;
+import io.envoyproxy.envoy.api.v2.auth.UpstreamTlsContext;
+import io.grpc.Server;
+import io.grpc.StatusRuntimeException;
+import io.grpc.stub.StreamObserver;
+import io.grpc.testing.GrpcCleanupRule;
+import io.grpc.testing.protobuf.SimpleRequest;
+import io.grpc.testing.protobuf.SimpleResponse;
+import io.grpc.testing.protobuf.SimpleServiceGrpc;
+import io.grpc.xds.internal.sds.SdsProtocolNegotiators;
+import io.grpc.xds.internal.sds.SecretVolumeSslContextProviderTest;
+import io.grpc.xds.internal.sds.XdsChannelBuilder;
+import io.grpc.xds.internal.sds.XdsServerBuilder;
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.util.Arrays;
+import javax.net.ssl.SSLHandshakeException;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.JUnit4;
+
+/**
+ * Unit tests for {@link XdsChannelBuilder} and {@link XdsServerBuilder} for plaintext/TLS/mTLS
+ * modes.
+ */
+@RunWith(JUnit4.class)
+public class XdsSdsClientServerTest {
+
+  /** Bad/untrusted server certs. */
+  private static final String BAD_SERVER_PEM_FILE = "badserver.pem";
+  private static final String BAD_SERVER_KEY_FILE = "badserver.key";
+
+  @Rule public final GrpcCleanupRule cleanupRule = new GrpcCleanupRule();
+  private Server server;
```

Comment:
this instance variable can be local variable.

Reply:
No, unless I declare a POJO just to return and use 2 values: the `Server` and `listenerWatcher`, since a function cannot return 2 values. Look at the flow in `mtlsClientServer_changeServerContext_expectException` calling `performMtlsTestAndGetListenerWatcher` and `buildClientAndTest`